## Panduan Penggunaan Model

Gunakanlah code di bawah ini untuk melakukan analisis terhadap proses preprocessing yang sudah dilakukan.
Keterangan:
1. Ubahlah dataset (csv file) sesuai dengan data kalian.
2. Gunakanlah dataset sebelum dan sesudah dilakukan proses preprocessing
3. Catat hasilnya
4. Bandingkan kedua hasil tersebut lalu buatlah kesimpulannya.

In [21]:
import numpy as np
import pandas as pd
import seaborn as sb
import os

##### Ubahlah dataset.csv sesuai dengan nama dataset kalian

In [22]:
dataframe = pd.read_csv('before_glcm_result.csv')
dataframe.head()

Unnamed: 0                                              path  contrast_0  \
0           0                        datasets/ayam\download.jpg   20.880116   
1           1  datasets/ayam\egg-white-food-protein-162712.jpeg  111.137362   
2           2                      datasets/ayam\images (1).jpg   49.763955   
3           3                          datasets/ayam\images.jpg   10.207475   
4           4             datasets/ayam\IMG_20220818_154051.jpg  212.539583   

   dissimilarity_0  homogeneity_0  energy_0  correlation_0     entropy_0  \
0         1.447426       0.782898  0.373523       0.992074 -8.327552e+07   
1         3.881204       0.669701  0.029680       0.974937 -6.290004e+07   
2         3.205836       0.570355  0.005943       0.995225 -3.821896e+07   
3         1.131587       0.722410  0.008221       0.991684 -6.367424e+07   
4         6.299969       0.420967  0.000707       0.975474 -3.143051e+07   

          asm_0  contrast_45  ...    entropy_95        asm_90  contrast_135  \
0  5.699516e-06    25.298101  ... -8.327552e+07  5.798912e-06     25.661130   
1  5.374321e-07   133.864852  ... -6.290004e+07  5.626249e-07    153.392434   
2  1.040406e-07    72.509189  ... -3.821896e+07  1.019906e-07    109.714633   
3  1.554063e-07    14.741515  ... -6.367424e+07  1.685736e-07     13.400261   
4  1.313005e-08   375.771642  ... -3.143051e+07  9.590911e-09    372.706067   

   dissimilarity_135  homogeneity_135  energy_135  correlation_135  \
0           1.634894         0.769432    0.371465         0.990283   
1           4.750650         0.638157    0.027062         0.965487   
2           4.950342         0.470013    0.004995         0.989470   
3           1.420838         0.676390    0.006677         0.989118   
4           9.175087         0.341886    0.000501         0.957054   

    entropy_135       asm_135  labels  
0 -8.327552e+07  5.668101e-06    ayam  
1 -6.290004e+07  4.900285e-07    ayam  
2 -3.821896e+07  8.745080e-08    ayam  
3 -6.367424e+07  1.262162e-07    ayam  
4 -3.143051e+07  9.302576e-09    ayam  

[5 rows x 31 columns]

In [23]:
# os.listdir(
#     'segmented-dataset'
# )

##### Ubahlah label sesuai dengan label kalian

In [24]:
clean_data  = dataframe.drop(columns=['Unnamed: 0', 'path'])
clean_data  = clean_data.replace ({
    'ayam': 0,
    'kalkun': 1,
    'puyuh': 2
})
clean_data.head()

contrast_0  dissimilarity_0  homogeneity_0  energy_0  correlation_0  \
0   20.880116         1.447426       0.782898  0.373523       0.992074   
1  111.137362         3.881204       0.669701  0.029680       0.974937   
2   49.763955         3.205836       0.570355  0.005943       0.995225   
3   10.207475         1.131587       0.722410  0.008221       0.991684   
4  212.539583         6.299969       0.420967  0.000707       0.975474   

      entropy_0         asm_0  contrast_45  dissimilarity_45  homogeneity_45  \
0 -8.327552e+07  5.699516e-06    25.298101          1.619946        0.770814   
1 -6.290004e+07  5.374321e-07   133.864852          4.417347        0.641840   
2 -3.821896e+07  1.040406e-07    72.509189          3.960477        0.515654   
3 -6.367424e+07  1.554063e-07    14.741515          1.514910        0.652074   
4 -3.143051e+07  1.313005e-08   375.771642          9.239569        0.332091   

   ...    entropy_95        asm_90  contrast_135  dissimilarity_135  \
0  ... -8.327552e+07  5.798912e-06     25.661130           1.634894   
1  ... -6.290004e+07  5.626249e-07    153.392434           4.750650   
2  ... -3.821896e+07  1.019906e-07    109.714633           4.950342   
3  ... -6.367424e+07  1.685736e-07     13.400261           1.420838   
4  ... -3.143051e+07  9.590911e-09    372.706067           9.175087   

   homogeneity_135  energy_135  correlation_135   entropy_135       asm_135  \
0         0.769432    0.371465         0.990283 -8.327552e+07  5.668101e-06   
1         0.638157    0.027062         0.965487 -6.290004e+07  4.900285e-07   
2         0.470013    0.004995         0.989470 -3.821896e+07  8.745080e-08   
3         0.676390    0.006677         0.989118 -6.367424e+07  1.262162e-07   
4         0.341886    0.000501         0.957054 -3.143051e+07  9.302576e-09   

   labels  
0       0  
1       0  
2       0  
3       0  
4       0  

[5 rows x 29 columns]

In [25]:
len(clean_data.columns)

29

In [26]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import MinMaxScaler
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix


x = clean_data.iloc[:,:-1]
y = clean_data.iloc[:,-1:]

x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2, random_state=0)

scaler = MinMaxScaler()
x_train_scalled = scaler.fit_transform(x_train)
x_test_scalled = scaler.transform(x_test)

##### Cek dokumentasi parameter pemanggilan fungsi pada list classifiers

In [27]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.svm import SVC
names = ["K-Nearest Neighbors", 
         "Linear SVM"]

classifiers = [
    KNeighborsClassifier(5),
    SVC(gamma=2, C=1),
    ]

#### Lakukanlah analisis hasil prediksi

In [28]:
for name, model in zip(names, classifiers):
  model.fit(x_train_scalled, y_train.values.ravel())
  y_pred_model = model.predict(x_test_scalled)
  print(f'## MODEL NAME : {name}')
  print(confusion_matrix(y_test, y_pred_model))
  print(classification_report(y_test, y_pred_model))
  print('\n')

## MODEL NAME : K-Nearest Neighbors
[[12  5  3]
 [ 3 14  3]
 [ 2  4 14]]
              precision    recall  f1-score   support

           0       0.71      0.60      0.65        20
           1       0.61      0.70      0.65        20
           2       0.70      0.70      0.70        20

    accuracy                           0.67        60
   macro avg       0.67      0.67      0.67        60
weighted avg       0.67      0.67      0.67        60



## MODEL NAME : Linear SVM
[[14  4  2]
 [ 3 11  6]
 [ 0  0 20]]
              precision    recall  f1-score   support

           0       0.82      0.70      0.76        20
           1       0.73      0.55      0.63        20
           2       0.71      1.00      0.83        20

    accuracy                           0.75        60
   macro avg       0.76      0.75      0.74        60
weighted avg       0.76      0.75      0.74        60





In [29]:
correlation = {}
label = clean_data['labels']
for col in clean_data.columns:
  correlation[col] = label.corr(clean_data[col])
sorted_correlation = sorted(correlation.items(), key=lambda x: x[1])
print(sorted_correlation)
len(sorted_correlation)

[('homogeneity_0', -0.508019917210892), ('homogeneity_45', -0.47996849305204997), ('homogeneity_135', -0.47597103706753807), ('homogeneity_90', -0.45065569444877773), ('correlation_135', -0.28106646600797813), ('correlation_45', -0.2557496982949805), ('correlation_0', -0.24795245496616258), ('asm_0', -0.20266587093144187), ('asm_90', -0.20184348263837698), ('asm_45', -0.20054343845390782), ('asm_135', -0.20052822668642192), ('energy_0', -0.1980718661466238), ('energy_90', -0.19752455405400865), ('energy_45', -0.1964475785745991), ('energy_135', -0.19642533470654808), ('correlation_90', -0.09418394854360665), ('entropy_0', 0.2313648622410407), ('entropy_45', 0.2313648622410407), ('entropy_95', 0.2313648622410407), ('entropy_135', 0.2313648622410407), ('contrast_90', 0.2981413787485418), ('contrast_0', 0.4284303192549676), ('contrast_45', 0.4529935006654953), ('dissimilarity_90', 0.45738612809552737), ('contrast_135', 0.47794541786828676), ('dissimilarity_0', 0.5564310727547226), ('dissi

29

In [30]:
positive_corr = {}
negative_corr = {}
for (k,v) in sorted_correlation:
  if(v>0):
    positive_corr[k] = v
  else:
    negative_corr[k] = v
len(positive_corr)

13

### PERBANDINGAN
Dalam membandingkan kedua model, K-Nearest Neighbors (KNN) dan Linear SVM, terdapat perbedaan sebagai berikut:

Akurasi: Akurasi berfungsi untuk mengukur sejauh mana model dapat memprediksi dengan benar. Pada model KNN, akurasi sebesar 0.60, sedangkan pada model Linear SVM, akurasi mencapai 0.73. Oleh karena itu, Linear SVM memiliki akurasi yang lebih tinggi daripada KNN.

Presisi dan Recall: Presisi mengukur sejauh mana hasil positif yang diprediksi oleh model adalah benar, sedangkan recall mengukur sejauh mana model dapat mengidentifikasi secara lengkap hasil positif yang sebenarnya. Model KNN memiliki presisi dan recall yang lebih rendah dibandingkan dengan model Linear SVM.

F1-Score: F1-score merupakan rata-rata antara presisi dan recall. Baik KNN maupun Linear SVM memiliki f1-score yang lumayan tinggi, tetapi model Linear SVM memiliki nilai f1-score yang sedikit lebih tinggi daripada KNN.


### KESIMPULAN
Berdasarkan perbandingan kedua model tersebut, dapat disimpulkan bahwa model Linear SVM memiliki hasil yang lebih baik daripada model K-Nearest Neighbors dalam memprediksi kelas pada dataset yang diberikan. Model Linear SVM memiliki akurasi yang lebih tinggi dan secara konsisten memberikan nilai presisi, recall, dan f1-score yang lebih baik untuk setiap kelas. Oleh karena itu, model Linear SVM direkomendasikan untuk digunakan dalam klasifikasi project ini.